src/mutable_list.jl

```diff
@@ -0,0 +1,217 @@
+mutable struct ListNode{T}
+    data::T
+    prev::ListNode{T}
+    next::ListNode{T}
+    function ListNode{T}() where T
+        node = new{T}()
+        node.prev = node
+        node.next = node
+        return node
+    end
+    function ListNode{T}(data) where T
+        node = new{T}(data)
+        return node
+    end
+end
+
+mutable struct MutableLinkedList{T}
+    len::Int
+    front::ListNode{T}
+    back::ListNode{T}
+    function MutableLinkedList{T}() where T
+        l = new{T}()
+        l.len = 0
+        l.front = ListNode{T}()
+        l.back = ListNode{T}()
+        l.front.next = l.back
+        l.back.prev = l.front
+        return l
+    end
+end
+
+MutableLinkedList() = MutableLinkedList{Any}()
+
+function MutableLinkedList{T}(elts...) where T
+    l = MutableLinkedList{T}()
+    for elt in elts
+        push!(l, elt)
+    end
+    return l
+end
+
+iterate(l::MutableLinkedList) = l.len == 0 ? nothing : (l.front.next.data, l.front.next.next)
+iterate(l::MutableLinkedList, n::ListNode) = n.next == n ? nothing : (n.data, n.next)
+
+isempty(l::MutableLinkedList) = l.len == 0
+length(l::MutableLinkedList) = l.len
+collect(l::MutableLinkedList{T}) where T = T[x for x in l]
+eltype(l::MutableLinkedList{T}) where T = T
+first(l::MutableLinkedList) = l.front.next.data
+last(l::MutableLinkedList) = l.back.prev.data
+
+function ==(l1::MutableLinkedList, l2::MutableLinkedList)
+    typeof(l1) == typeof(l2) || return false
```

Comment:
This is actually a conditional on the element type, since we know that the wrappers are the same. So you could do
```julia
==(l1::MutableLinkedList{T}, l2::MutableLinkedList{S}) where {T,S} = false

function ==(l1::MutableLinkedList{T}, l2::MutableLinkedList{T}) where T
    # the rest of the checking
end
```